### DISCUSSION TITLE
Getting Started

### DISCUSSION THREAD

**Discussion by shubhendumadhukar** [fauxauldrich/camouflage#22] [General]
This discussion is for the new comers who would like to use Camouflage for their mocking needs. This discussion would focus strictly on providing help with onboarding.

**Comment by jimsheen** [comment]
Hi there, thanks for the package. Is there support for importing JSON files like mockserver package does? I found that extremely useful

**Reply by shubhendumadhukar** [reply]
Sure. It is supported, but implemented a bit differently. You can create your mock file as `${HTTP_METHOD}.mock` and use `file` handlebar to serve data from json file. Content of your mock file could look something like following:
```
HTTP/1.1 200 OK
Content-Type: application/json

{{file path="/Users/user/Desktop/test.json"}}
```

Let me know if this works for you.

**Reply by jimsheen** [reply]
Is it possible to import js files ?

**Reply by jsapes** [reply]
@shubhendumadhukar I'm interested in include a js file. There is a guide how to do it?

**Reply by shubhendumadhukar** [reply]
@jsapes, You can probably import/require a js file using the `code` helper.

I will get back to you with a sample mock file.

**Reply by shubhendumadhukar** [reply]
Modifying the existing example mock for `code`  helper a bit, gives me the following sample mock. (You'd need to enable injection in config.yml for this to work)
```js
HTTP/1.1 200 OK
Content-Type: application/json

{{#code}}
(()=>{
    const {getRandomNumberInRange} = require('/home/shubhendu/Desktop/sample-project/rand.js')
    const name = request.query.name;
    const phone = getRandomNumberInRange(1000000000, 9999999999);
    logger.info(phone);
    return {
        status: 201,
        headers: {
            'X-Requested-By': name
        },
        body: `{
            "greeting": "Hello ${name}",
            "phone": ${phone}
        }`
    };
})();
{{/code}}
```
The paths are a bit tricky to get the hang of so I would use absolute path while requiring an external js file to avoid any errors.

**rand.js**
```js
function getRandomNumberInRange(min, max) {
    return Math.round(Math.random() * (max - min) + min);
}

module.exports = { getRandomNumberInRange }
```
Let me know if this works for you.

**Reply by jsapes** [reply]
Hi @shubhendumadhukar Yes it works! I will put my example here getting the path dynamically.

Camouflage dir
─ package.json
─ camouflage.config.yml
─ mocks/myServer/my/path/ GET.mock & getFunction.js

   
 
   ```
 File: GET.mock
  
  
 HTTP/1.1 200 OK
Content-Type: application/json

{{#code}}
(()=>{
    const path = require("path")
    const myPath = path.join(process.cwd(), "mocks", request.url, "getFunction.js")
    const { getFunction } = require(myPath)
    return myFunction(request)
})();
{{/code}}
  
   ```   

  ``` js     
  File:  getFunction.js
                      
  const getFunction = (request) => {
  // myCode
    return {
      status: 201,
      body: { ...myBody  },
    };
  };

module.exports = { getFunction };

  ```

**Comment by voronovmaksim** [comment]
Hi. thanks for the tool. 
I need to mock Websocket. I want camouflage to push data from a file(example csv, pg etc) line by line. In other word, i want to prepare a queue of message:
```
message 1
message 2
message 3
message 4
```

And get the message in the same order:
```
new WebSocket("ws://127.0.0.1:8082/hello/world").onmessage = function(event) {
        console.log("socket data:" + event.data);
    };
```
Result in console:
```
socket data: message 1
socket data: message 2
socket data: message 3
socket data: message 4
```

Is it implemented? I didin't find any information about it.

Thanks.

**Reply by shubhendumadhukar** [reply]
Thanks @YasuroNN for an interesting use case. Camouflage doesn't support it at this time, but I'd be happy to implement the functionality. Watch #146 for updates.

Note that the issue created, is for csv data source. If you are open to using postgres, the mock you described, can easily be created with `pg` helper. Let me know if you need help with that. Meanwhile I'll be implementing the same with CSV data sources as well.

**Reply by voronovmaksim** [reply]
Thanks a lot. It will be great if this one is implemented for cvs. I am open to using postgres, but I don't understand how it can be reach. I would appreciate it if you help me. As I understand, pg helper allows executing a query to postgres in just stateless mode(i cannot use a variable in js to iterate through result of query)

**Reply by shubhendumadhukar** [reply]
Apologies for the delay in response. I am working on creating a sample mock for your use case with both `pg` and `csv` which you can use as reference.

**Reply by voronovmaksim** [reply]
Ok, thanks. Let me know please here when it Will be ready

**Reply by voronovmaksim** [reply]
@shubhendumadhukar,  if you are developing new feature, it will be nice if you allow to set frequency of mock websocket message also.  I think, This Will alow mock server to be more flexible)

**Comment by FlemmingBehrend** [comment]
Hi

In the [documentation](https://testinggospels.github.io/camouflage/request-matching/) is says that it is possible to make responses depending on an Authorization header is set etc.

I have tried the following but it always returns the `no headers` response even when I have set a test header.

```
{{#if request.headers.test}}
HTTP/1.1 200 OK
Content-Type: application/json; charset=utf-8
Access-Control-Allow-Origin: *
Response-Delay: 100

{
   "Random": "headers"
}
{{else}}
HTTP/1.1 200 OK
Content-Type: application/json; charset=utf-8
Access-Control-Allow-Origin: *
Response-Delay: 500

{
   "Random": "no headers"
}
{{/if}}
```

Is it supported at the moment?

**Reply by shubhendumadhukar** [reply]
There is a bug currently in camouflage@0.8.1 which prevents this. This should be fixed in 0.8.2. Please refer to #181 for a detailed explaination.

**Reply by shubhendumadhukar** [reply]
Documentation updated for [Request Matching using headers](https://testinggospels.github.io/camouflage/request-matching/#request-matching-using-headers)

**Comment by nimi18** [comment]
Hi team,

I am facing below issue while running the mock server 

![image](https://user-images.githubusercontent.com/56684916/203309229-50f6fc9a-a8dc-4787-8802-a943f97fed63.png)

I am trying to setup react store sdk on windows 10 machine. The sdk works fine on Ubuntu OS but getting error for workspace in windows.

Can someone please explain the issue and let me know asap.

**Reply by shubhendumadhukar** [reply]
Hi @nimi18,

From the screenshot shared, it's not clear what is the issue you are facing. Or if it even is related to camouflage. It seems a package `@hcl-commerce-store-sdk/react-component` is missing, which doesn't seem to be related to camouflage. Some more context would be helpful.

**Comment by nimi18** [comment]
Hi Shubhendu,

I got hold of this discussion page through one of the HCL commerce links.

By any chance are you aware of the HCL Commerce React Store SDK?

Please let me know so that we can take the conversation ahead.

Warm Regards,
Nimita Malhotra

Get Outlook for iOS<https://aka.ms/o0ukef>
________________________________

**Reply by shubhendumadhukar** [reply]
Unfortunately no, the issue you're facing doesn't seem to be related to this project.

**Comment by nimi18** [comment]
Ohkay.

Anyways, thanks for replying at such short notice.



Get Outlook for iOS<https://aka.ms/o0ukef>
________________________________

**Comment by dnafication** [comment]
Hi all,

I am assessing if camouflage will be a good fit for my usecase of replacing a REST API. I have noticed that it supports running queries against `pg` [Link](https://testinggospels.github.io/camouflage/external-data-source/). Does camouflage allow us to push (insert) data to `pg`?

**Reply by shubhendumadhukar** [reply]
Admittedly it might not be very readable with larger queries but passing raw insert query does allow inserts to the database. Tested following snippet and it works just fine.

```
HTTP/1.1 201 OK
Content-Type: application/json

{{#pg query='INSERT INTO test_table (name) VALUES (\'shubhendu\')'}}
(()=> {
    
    return {
        status: 201,
        headers: {
            'Content-Type': 'application/json'
        },
        body: JSON.stringify(result)
    };
})();
{{/pg}}
```

**Comment by mcvendrell** [comment]
Hi.

Maybe it is obvious, but I still don't know how to do it, looking at mocks folder examples: I have a classic JSON data of the pets folder example

```
[
  { "id": 1, "name": "Rabbit" },
  { "id": 2, "name": "Dog" },
  { "id": 3, "name": "Cat" },
  { "id": 4, "name": "Bird" }
]
```

How can I filter data in the http call? Something like http://localhost:3000/pets/1 or http://localhost:3000/pets?id=1
Or how can I filter by some field? Let's say http://localhost:3000/pets?name=Cat

**Reply by mcvendrell** [reply]
anyone? @shubhendumadhukar maybe?

**Reply by shubhendumadhukar** [reply]
Hello,

Apologies for the delay in response, I am no longer actively maintaining this project.

Regarding your query, you could do it by using one of the following 2 approaches.
1. Use a `code` helper, your mock would look something like this
```javascript
HTTP/1.1 200 OK
Content-Type: application/json

{{#code}}
(()=>{
    const id = parseInt(request.query.id);
    const data = [
        { "id": 1, "name": "Rabbit" },
        { "id": 2, "name": "Dog" },
        { "id": 3, "name": "Cat" },
        { "id": 4, "name": "Bird" }
    ]
    const response = data.filter((el)=>el.id === id)
    return {
        status: 200,
        body: JSON.stringify(response[0])
    };
})();
{{/code}}
```
2. Second approach would be to use the csv helper
a. Create a csv file, data.csv
```
id;name
1;Rabbit
2;Dog
3;Cat
4;Bird
```
b. Create a mock as shown below:
```javascript
HTTP/1.1 200 OK
Content-Type: application/json

{{#csv src="./data.csv" key="id" value=(capture from='query' key='id') }}
(()=> {
    return {
        status: 200,
        body: JSON.stringify(result[0])
    };
})();
{{/csv}}
```

Edit:
1. Note that you'd have to enable injection via your config yaml, for either of these helpers (i.e. code/csv) to work.
2. Similar approach could be used in case you want to use path parameters, you'd have to use regex to capture the value of id.

More info on csv helper [here](https://testinggospels.github.io/camouflage/external-data-source/#csv), and code helper [here](https://testinggospels.github.io/camouflage/handlebars/#code)

**Reply by mcvendrell** [reply]
Thank you @shubhendumadhukar. So no "direct" method with only data, but programmatically.
You are not actively mantaining but still "hearing to it"? Or it is dead? I think it's a very good project, so "advanced" compared to others.
Can I create some docs and do a PR?

**Reply by shubhendumadhukar** [reply]
>So no "direct" method with only data, but programmatically.

Unfortunately no. Using `csv + capture` as shown in 2nd example, would be simpler. You can get away with lesser code since it does the filtering for you, and you only have to return the response. Using `code`, you have more control. But in any case, it has to be done programmatically.

>You are not actively mantaining but still "hearing to it"? Or it is dead? I think it's a very good project, so "advanced" compared to others.

This project was initially created to help the needs of a small organization, however it turned quickly into an unmaintainable mess. I am currently re-creating the project from scratch with maintainability and more importantly, extensibility in mind (protocol specific modules, @camoflage/http, @camoflage/grpc and @camoflage/helpers and so on, download only what you need, smaller and less confusing configs, hooks, etc) but it's still in extremely early stages. I'll probably create an "issue" with more details when I am closer to a release. I'll be using this project meanwhile, only for feedbacks and features I could implement in the new project.

**Reply by mcvendrell** [reply]
Oh, what a great notice. I feel this project as an amazing free tool for API mock. I was using "json-server", but the licensing method prevents its use in companies.

By the way, I will enhance your readme page, because it is not clear enough how to reach de doc page (that i discover right now) and will add this classic example of "filter", because is the common use people search for mocks. I will send you a PR, if you want to include it.

Thank you for this work.

**Reply by mcvendrell** [reply]
I tested your solution and found two problems I didn't expected: getting errors about *unknown query parameter* or "JSON format problems". The first came from schema validation (I disabled it) and the second from "injection" option (seems should be "true" in order to use Handlebars).

After it, all works fine, so I did some documentation and put it in the README for future references and for newbies like me. Did a PR, if you want to publish @shubhendumadhukar .

Thanks for all.

**Comment by joseph-hanvy-sophos** [comment]
I'm getting module issues (`Error: Cannot find module 'soap'`), just trying to `camouflage init` or running it (can do `--version` though):

```
camouflage init

internal/modules/cjs/loader.js:905
  throw err;
  ^

Error: Cannot find module 'soap'
Require stack:
- /home/.../node_modules/express-soap/lib/express-soap.js
- /home/.../node_modules/express-soap/index.js
- /home/.../node_modules/camouflage-server/dist/protocols/Soap.js
- /home/.../node_modules/camouflage-server/dist/protocols/index.js
- /home/.../node_modules/camouflage-server/dist/index.js
- /home/.../node_modules/camouflage-server/bin/camouflage.js
    at Function.Module._resolveFilename (internal/modules/cjs/loader.js:902:15)
    at Function.Module._load (internal/modules/cjs/loader.js:746:27)
    at Module.require (internal/modules/cjs/loader.js:974:19)
    at require (internal/modules/cjs/helpers.js:101:18)
    at Object.<anonymous> (/home/.../node_modules/express-soap/lib/express-soap.js:4:20)
    at Module._compile (internal/modules/cjs/loader.js:1085:14)
    at Object.Module._extensions..js (internal/modules/cjs/loader.js:1114:10)
    at Module.load (internal/modules/cjs/loader.js:950:32)
    at Function.Module._load (internal/modules/cjs/loader.js:790:12)
    at Module.require (internal/modules/cjs/loader.js:974:19) {
  code: 'MODULE_NOT_FOUND',
  requireStack: [
    '/home/.../node_modules/express-soap/lib/express-soap.js',
    '/home/.../node_modules/express-soap/index.js',
    '/home/.../node_modules/camouflage-server/dist/protocols/Soap.js',
    '/home/.../node_modules/camouflage-server/dist/protocols/index.js',
    '/home/.../node_modules/camouflage-server/dist/index.js',
    '/home/.../node_modules/camouflage-server/bin/camouflage.js'
  ]
}
```

I'm using node 14, are there any other requirements I'm missing?

**Reply by mcvendrell** [reply]
Did you install it globally? `npm install -g camouflage-server`

Also, for me is working well, but I'm using node v20

**Reply by joseph-hanvy-sophos** [reply]
Doesn't work globally either with 14 with the same error. Switched to 18 and it seems to function fine - so definitely is the node version. Annoying we're tied to such an old version!

**Reply by mcvendrell** [reply]
In the readme, it is not clear which node version to use. Well, mathematically it is clear, but for user readability, it is not. See:

> you'd need to install NodeJS (>v14) first

Node GREATER than v14.
